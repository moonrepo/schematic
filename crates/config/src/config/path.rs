use std::fmt::{self, Display};

/// Represents all the different forms a path is composed of.
#[derive(Clone, Debug)]
pub enum PathSegment {
    /// List index: `[0]`
    Index(usize),
    /// Map key: `name.`
    Key(String),
    /// Enum variant: `name.`
    Variant(String),
    /// Unknown segment: `?`
    Unknown,
}

/// Represents the path from the struct root to nested a field or field value.
#[derive(Clone, Debug, Default)]
pub struct Path {
    /// List of path segments.
    segments: Vec<PathSegment>,
}

impl Path {
    /// Create a new instance with the provided [`PathSegment`]s.
    pub fn new(segments: Vec<PathSegment>) -> Self {
        Self { segments }
    }

    /// Create a new instance and append the provided [`PathSegment`]
    /// to the end of the current path.
    pub fn join(&self, segment: PathSegment) -> Self {
        let mut path = self.clone();
        path.segments.push(segment);
        path
    }

    /// Create a new instance and append an `Index` [`PathSegment`]
    /// to the end of the current path.
    pub fn join_index(&self, index: usize) -> Self {
        self.join(PathSegment::Index(index))
    }

    /// Create a new instance and append an `Key` [`PathSegment`]
    /// to the end of the current path.
    pub fn join_key(&self, key: &str) -> Self {
        self.join(PathSegment::Key(key.to_owned()))
    }

    /// Create a new instance and append another [`Path`]
    /// to the end of the current path.
    pub fn join_path(&self, other: &Self) -> Self {
        let mut path = self.clone();
        path.segments.extend(other.segments.clone());
        path
    }

    /// Create a new instance and append an `Variant` [`PathSegment`]
    /// to the end of the current path.
    pub fn join_variant(&self, variant: &str) -> Self {
        self.join(PathSegment::Variant(variant.to_owned()))
    }
}

impl Display for Path {
    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        if self.segments.is_empty() {
            return formatter.write_str(".");
        }

        let mut separator = "";

        for segment in &self.segments {
            match segment {
                PathSegment::Index(index) => {
                    write!(formatter, "[{}]", index)?;
                }
                PathSegment::Key(key) | PathSegment::Variant(key) => {
                    write!(formatter, "{}{}", separator, key)?;
                }
                PathSegment::Unknown => {
                    write!(formatter, "{}?", separator)?;
                }
            }

            separator = ".";
        }

        Ok(())
    }
}
