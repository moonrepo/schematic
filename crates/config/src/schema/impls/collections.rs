// use crate::config::ConfigSchema;
// use crate::schema::types::*;

// impl<T: ConfigSchema> ConfigSchema for Vec<T> {
//     fn generate_schema() -> Schema {
//         let inner = T::generate_schema();

//         if let Schema::Type {
//             ref mut nullable, ..
//         } = schema
//         {
//             *nullable = true;
//         }

//         schema
//     }
// }
