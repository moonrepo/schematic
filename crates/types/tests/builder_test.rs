use schematic_types::*;
use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
use std::net::Ipv4Addr;
use std::path::{Path, PathBuf};
use std::time::Duration;

fn test_builder<T: Schematic>() -> Schema {
    SchemaBuilder::build_root::<T>()
}

pub struct Named {
    pub field: bool,
}

impl Schematic for Named {
    fn schema_name() -> Option<String> {
        Some("Named".into())
    }

    fn build_schema(mut schema: SchemaBuilder) -> Schema {
        schema.structure(StructType::new([("field".into(), schema.infer::<bool>())]))
    }
}

#[test]
fn primitives() {
    assert_eq!(test_builder::<()>().ty, SchemaType::Null);

    assert_eq!(
        test_builder::<bool>().ty,
        SchemaType::Boolean(Box::default())
    );

    assert_eq!(
        test_builder::<&bool>().ty,
        SchemaType::Boolean(Box::default())
    );

    assert_eq!(
        test_builder::<&mut bool>().ty,
        SchemaType::Boolean(Box::default())
    );

    assert_eq!(
        test_builder::<Box<bool>>().ty,
        SchemaType::Boolean(Box::default())
    );

    assert_eq!(
        test_builder::<Option<bool>>().ty,
        SchemaType::Union(Box::new(UnionType::new_any(vec![
            SchemaType::Boolean(Box::default()),
            SchemaType::Null
        ])))
    );
}

#[test]
fn arrays() {
    assert_eq!(
        test_builder::<Vec<String>>().ty,
        SchemaType::Array(Box::new(ArrayType::new(SchemaType::String(Box::default()))))
    );

    assert_eq!(
        test_builder::<&[String]>().ty,
        SchemaType::Array(Box::new(ArrayType::new(SchemaType::String(Box::default()))))
    );

    assert_eq!(
        test_builder::<[String; 3]>().ty,
        SchemaType::Array(Box::new(ArrayType {
            items_type: Box::new(Schema::string(StringType::default())),
            max_length: Some(3),
            min_length: Some(3),
            ..ArrayType::default()
        }))
    );

    assert_eq!(
        test_builder::<HashSet<String>>().ty,
        SchemaType::Array(Box::new(ArrayType {
            items_type: Box::new(Schema::string(StringType::default())),
            unique: Some(true),
            ..ArrayType::default()
        }))
    );

    assert_eq!(
        test_builder::<BTreeSet<String>>().ty,
        SchemaType::Array(Box::new(ArrayType {
            items_type: Box::new(Schema::string(StringType::default())),
            unique: Some(true),
            ..ArrayType::default()
        }))
    );
}

#[test]
fn integers() {
    assert_eq!(
        test_builder::<u8>().ty,
        SchemaType::Integer(Box::new(IntegerType::new_kind(IntegerKind::U8)))
    );

    assert_eq!(
        test_builder::<i32>().ty,
        SchemaType::Integer(Box::new(IntegerType::new_kind(IntegerKind::I32)))
    );
}

#[test]
fn floats() {
    assert_eq!(
        test_builder::<f32>().ty,
        SchemaType::Float(Box::new(FloatType::new_kind(FloatKind::F32)))
    );

    assert_eq!(
        test_builder::<f64>().ty,
        SchemaType::Float(Box::new(FloatType::new_kind(FloatKind::F64)))
    );
}

#[test]
fn objects() {
    assert_eq!(
        test_builder::<HashMap<String, Named>>().ty,
        SchemaType::Object(Box::new(ObjectType::new(
            Schema::string(StringType::default()),
            test_builder::<Named>(),
        )))
    );

    assert_eq!(
        test_builder::<BTreeMap<u128, Named>>().ty,
        SchemaType::Object(Box::new(ObjectType::new(
            SchemaType::Integer(Box::new(IntegerType::new_kind(IntegerKind::U128))),
            test_builder::<Named>(),
        )))
    );
}

#[test]
fn strings() {
    assert_eq!(
        test_builder::<char>().ty,
        SchemaType::String(Box::new(StringType {
            max_length: Some(1),
            min_length: Some(1),
            ..StringType::default()
        }))
    );

    assert_eq!(
        test_builder::<&str>().ty,
        SchemaType::String(Box::default())
    );

    assert_eq!(
        test_builder::<String>().ty,
        SchemaType::String(Box::default())
    );

    assert_eq!(
        test_builder::<&Path>().ty,
        SchemaType::String(Box::new(StringType {
            format: Some("path".into()),
            ..StringType::default()
        }))
    );

    assert_eq!(
        test_builder::<PathBuf>().ty,
        SchemaType::String(Box::new(StringType {
            format: Some("path".into()),
            ..StringType::default()
        }))
    );

    assert_eq!(
        test_builder::<Ipv4Addr>().ty,
        SchemaType::String(Box::new(StringType {
            format: Some("ipv4".into()),
            ..StringType::default()
        }))
    );

    assert_eq!(
        test_builder::<Duration>().ty,
        SchemaType::String(Box::new(StringType {
            format: Some("duration".into()),
            ..StringType::default()
        }))
    );
}

#[test]
fn tuples() {
    assert_eq!(
        test_builder::<(bool, i16, f32, String)>().ty,
        SchemaType::Tuple(Box::new(TupleType::new(vec![
            SchemaType::Boolean(Box::default()),
            SchemaType::Integer(Box::new(IntegerType::new_kind(IntegerKind::I16))),
            SchemaType::Float(Box::new(FloatType::new_kind(FloatKind::F32))),
            SchemaType::String(Box::default())
        ])))
    );
}

pub struct Cycle {
    pub values: HashMap<String, Cycle>,
}

impl Schematic for Cycle {
    fn schema_name() -> Option<String> {
        Some("Cycle".into())
    }

    fn build_schema(mut schema: SchemaBuilder) -> Schema {
        schema.structure(StructType::new([(
            "values".into(),
            schema.infer::<HashMap<String, Cycle>>(),
        )]))
    }
}

#[test]
fn supports_cycles() {
    assert_eq!(
        test_builder::<Cycle>().ty,
        SchemaType::Struct(Box::new(StructType {
            fields: BTreeMap::from_iter([(
                "values".into(),
                Box::new(Schema::object(ObjectType::new(
                    Schema::string(StringType::default()),
                    SchemaType::Reference("Cycle".into()),
                )))
            ),]),
            partial: false,
            required: None
        }))
    );
}
