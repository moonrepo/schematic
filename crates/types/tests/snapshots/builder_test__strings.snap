---
source: crates/types/tests/builder_test.rs
expression: "& input"
---
{
  "ty": {
    "type": "String",
    "max_length": 1,
    "min_length": 1
  }
}
